### PR TITLE
IMAP Connector

### PR DESCRIPTION
**Pull request by crocker**
# Tl;dr
Added IMAP Connector

# Details
The IMAP connector downloads and parses email messages from an IMAP server.

# How to verify
- Follow the build instructions
- Upload the connector to the Fusion Blob Store
- Configure a new IMAP data source
- Start the data source job and verify that messages are being downloaded with the following fields: From, To, Cc, Reply To, Subject, Body

### CODE REVIEW

File: plugins/java/connectors/imap-connector/src/main/java/com/lucidworks/fusion/connector/plugin/client/ImapClient.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+  public List<Email> getMessages(String folderName) throws MessagingException, IOException {
+    validConnection();
+
+    Folder folder = store.getFolder(folderName);
+    folder.open(Folder.READ_ONLY);
+    Message[] messages = folder.getMessages();
+    List<Email> emails = new ArrayList<>();
+    for(Message message : messages) {
+      emails.add(toEmail(message));
```
Comment (wsolaligue) [inline plugins/java/connectors/imap-connector/src/main/java/com/lucidworks/fusion/connector/plugin/client/ImapClient.java]: Please use lambda here to iterate the messages, and the method `.map` to convert the `Message` to `Email`, then use `.collect(Collectors.toList())` to set the emails

File: plugins/java/connectors/imap-connector/src/main/java/com/lucidworks/fusion/connector/plugin/client/ImapClient.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    Folder folder = store.getFolder(folderName);
+    folder.open(Folder.READ_ONLY);
+
+    Message[] messages = folder.search(
+        new FlagTerm(new Flags(Flags.Flag.SEEN), false));
+
+    List<Email> emails = new ArrayList<>();
+    for(Message message : messages) {
+      emails.add(toEmail(message));
```
Comment (wsolaligue) [inline plugins/java/connectors/imap-connector/src/main/java/com/lucidworks/fusion/connector/plugin/client/ImapClient.java]: Same as above here

File: plugins/java/connectors/imap-connector/src/main/java/com/lucidworks/fusion/connector/plugin/client/ImapClient.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    }
+  }
+
+
+  private List<String> addressToString(Address[] addresses) {
+    List<String> addrs = new ArrayList<String>();
+
+    if(addresses != null) {
+      for (Address address : addresses) {
+        addrs.add(address.toString().replaceAll("\"", ""));
```
Comment (wsolaligue) [inline plugins/java/connectors/imap-connector/src/main/java/com/lucidworks/fusion/connector/plugin/client/ImapClient.java]: Same as above, but use the method `map` to convert the `Address` to `String`

File: plugins/java/connectors/imap-connector/src/main/java/com/lucidworks/fusion/connector/plugin/ImapFetcher.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  @Override
+  public FetchResult fetch(FetchContext fetchContext) {
+    try {
+      for (Email message : imapClient.getUnreadMessages(this.config.getProperties().getFolder())) {
+        Map<String, Object> data = new HashMap<>();
+
+
+        data.put("id", message.getId());
```
Comment (wsolaligue) [inline plugins/java/connectors/imap-connector/src/main/java/com/lucidworks/fusion/connector/plugin/ImapFetcher.java]: Please use Constants for `id`, `from`, `reply_to`, `to`, etc. You might create a class named `ImapConstants` and put these within it.

File: plugins/java/connectors/imap-connector/src/main/java/com/lucidworks/fusion/connector/plugin/ImapFetcher.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        data.put("from", message.getFrom());
+        data.put("reply_to", message.getReplyTo());
+        data.put("to", message.getTo());
+        data.put("cc", message.getCc());
+        data.put("date", message.getDate().toInstant().toEpochMilli());
+        data.put("subject", message.getSubject());
+        data.put("body", message.getBody());
+
+        for(Map.Entry<String, Object> entry: data.entrySet()) {
+          if(entry.getValue() == null) data.remove(entry.getKey());
```
Comment (wsolaligue) [inline plugins/java/connectors/imap-connector/src/main/java/com/lucidworks/fusion/connector/plugin/ImapFetcher.java]: Format here please, put `data.remove(entry.getKey());` within curly brackets

**Comment by acesar** [comment]
Closing in favor of https://github.com/lucidworks/connectors-sdk-resources/pull/20

**Pull request closed without merging**